### ISSUE TITLE
push to upstream terraform repo

### ISSUE THREAD

**Issue by philips** [coreos/terraform-provider-ct#2]
Blockers:

- fuze is not the right name

**Comment by s-urbaniak** [comment]
It seems ignition configuration has been upstreamed already: https://github.com/hashicorp/terraform/blob/master/builtin/providers/ignition

https://www.terraform.io/docs/providers/ignition/r/user.html

**Comment by philips** [comment]
@bison can we shut down this repo then?

**Comment by bison** [comment]
@philips Seeing that there's now Ignition support actually encourages me to figure out naming and try to get this upstream ASAP.

I think there's still room for using templated Fuze / CT YAML configs that are converted to Ignition over building configs with potentially dozens of individual resources.

**Comment by philips** [comment]
Can you outline some of those use cases?

**Comment by crawford** [comment]
This is exactly where terraform excels. You write a mostly environment-agnostic config and let terraform figure out the details of that particular environment. Container Linux Configs are designed to be treated the same way - the user writes one config for all of their environments. Terraform can then use CT to make the customizations needed.

For example, given the following Container Linux Config:

```yaml
etcd:
  version:                     "v3.0.15"
  advertise_client_urls:       "{public.IPv4}:2379"
  initial_advertise_peer_urls: "{private.IPv4}:2380"
  listen_client_urls:          "http://0.0.0.0:2379"
  listen_peer_urls:            "http://{private.IPv4}:2380"
  initial_cluster:             "%m=http://{private.IPv4}:2380"
```

CT will use `COREOS_DIGITALOCEAN_IPV4_PUBLIC_0` and `COREOS_DIGITALOCEAN_IPV4_PRIVATE_0` for `public.IPv4` and `private.IPv4` on DigitalOcean, respectively. On EC2 though, it uses `COREOS_EC2_IPV4_PUBLIC` and `COREOS_EC2_IPV4_LOCAL`.

If the config looked like the following (IPv6 vs IPv4):

```yaml
etcd:
  version:                     "v3.0.15"
  advertise_client_urls:       "{public.IPv6}:2379"
  initial_advertise_peer_urls: "{private.IPv6}:2380"
  listen_client_urls:          "http://0.0.0.0:2379"
  listen_peer_urls:            "http://{private.IPv6}:2380"
  initial_cluster:             "%m=http://{private.IPv6}:2380"
```

CT would similarly use `COREOS_DIGITALOCEAN_IPV6_PUBLIC_0` and `COREOS_DIGITALOCEAN_IPV6_PRIVATE_0` for DigitalOcean. If a user attempted to use this on EC2 though, terraform would fail, informing the user that we can't get IPv6 addresses on EC2.

By using Container Linux Configs instead of Ignition Configs, we can hide all of this complexity from the end user, yet still make it available to those who need it.

**Comment by bison** [comment]
The use case is ultimately the same as the existing Ignition provider. I just think some people would prefer to manage a config in a single YAML document rather than a collection of individual Terraform resources.

**Comment by saracen** [comment]
@crawford Regarding `COREOS_DIGITALOCEAN_IPV4_PUBLIC_0` vs. `COREOS_EC2_IPV4_PUBLIC` vs. `PUBLIC_IPV4`, why was it decided that https://github.com/coreos/coreos-metadata wouldn't just expose common attributes under identical names, regardless of the environment?

**Comment by crawford** [comment]
Let's take Azure as an example. A machine will typically have a virtual, dynamic, and public address (respectively, VIP, DIP, and PIP) when created via the ARM APIs. However a machine created through the ASM APIs will only have a DIP and VIP. Both the PIP and VIP may be used to address the machine on the internet, but the PIP is preferred (the VIP is the address of a load balancer). In this case, which address do we use as the "public IP"? Pretty difficult to answer and even more difficult for an end-user to remember.

This is why we make ct figure it out and let the user double check the decision.

**Comment by saracen** [comment]
Having `coreos-metadata` also make that decision would work though, right? If it were to expose both the `COREOS_`**ENV**`_*` version, as well as make assumptions on what `PUBLIC_IPV4` should be (similar to ct).

I ask because I was looking to add a new metadata variable to coreos-metadata (gce zone and aws placement/availability), and I was going to choose an environment-agnostic name but then realized that nothing else had been done this way. Then I stumbled across this project. It's mildly inconvenient to keep the two projects synchronized with features that some may only add to `coreos-metadata` and not `terraform-provider-ct`. But, it is only a very mild inconvenience :)

**Comment by crawford** [comment]
Exposing generic names via `coreos-metadata` is an interesting idea. It's not without its own set of problems though. Let's take Azure again as an example (not picking on them, just intimately familiar with the platform). The PIPs are a newer concept, which cannot be discovered from within the machine yet. So let's go with VIPs in the meantime for `PUBLIC_IPV4`. Fast-forward a bit and now PIPs are discoverable from within the machine and the ASM APIs have been removed (effectively killing off the use of VIPs). We cannot safely change the `PUBLIC_IPV4` from the VIP to the PIP since it will break everyone who depended on that behavior. We are stuck forever.

Every environment has similar problems. I'm of the pretty strong opinion that magic is for children's birthdays and everything should be explicitly specified. But this is only after fighting with these exact problems for years.

**Comment by saracen** [comment]
I see. That makes a lot of sense. Thank you for taking the time to explain.

**Comment by dghubble** [comment]
Terraform modules don't need to be contributed to a single upstream repository. They're actively being split out per https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/

**Comment by dghubble** [comment]
Upstream docs on how to load custom plugins is forthcoming, but this makes it clear the need to merge into the terraform repo is gone:

"Provider plugins are no longer included with the main Terraform binary distribution. Instead, they are distributed separately and are fetched and installed on-demand by the terraform init command. This new approach allows users to upgrade individual providers, use custom or patched providers, and remove providers in isolation."

Closing

**Comment by MonaHBO** [comment]
is there a plan to check it in the repo  https://github.com/terraform-providers, so that it can be pulled by terraform init

**Comment by dghubble** [comment]
I support it and can champion/maintain it. Anyone from Hashicorp with push access?

**Comment by dghubble** [comment]
See #21
